src/math/interpolator.hh

```diff
@@ -0,0 +1,1649 @@
+                         start_grid_fineness, true) {}
+      Interpolator(double x1, double x2, double error_bound)
+          : Interpolator(x1, x2, error_bound, 10000000, 5, true) {}
+
+      /**
+       * Constructor to initialize interpolator from grid.
+       * @param grid in the range [x1,x2]
+       */
+      explicit Interpolator(Vector_t grid) : Interpolator(grid, true) {}
+
+      int get_grid_fineness() { return this->grid_fineness; }
+      int get_grid_size() { return this->grid.size(); }
+      Vector_Ref get_grid_ref() { return Vector_Ref(this->grid); }
+
+      // Should be implemented, but is not made virtual for performance
+      // virtual Ω interpolate(double x);
+      // virtual Ω interpolate_derivative(double x);
+      // virtual Vector_Ref get_grid_ref() = 0;
+      // virtual Ω  get_evaluated_grid_ref() = 0;
+      // virtual int get_grid_size() = 0;
+
+     protected:
+      Interpolator(double x1, double x2, double error_bound,
+                   int max_grid_points, int start_grid_fineness, bool)
+          : x1{x1}, x2{x2}, error_bound{error_bound},
+            max_grid_points{max_grid_points},
+            grid_fineness{start_grid_fineness},
+            intp_method{InterpolationMethod()}, grid_rational{GridRational()},
+            search_method{SearchMethod()} {
+        if (x2 < x1) {
+          throw std::logic_error("x2 must be greater x1");
+        }
+        if (error_bound <= 0) {
+          throw std::logic_error("Error bound must be > 0");
+        }
+        if (start_grid_fineness < 1) {
+          throw std::logic_error("Starting grid fineness must be at least 1.");
+        }
+        if (max_grid_points < 2) {
+          throw std::logic_error(
+              "Maximal number of grid points must be at least 2.");
+        }
+      };
+
+      Interpolator(Vector_t grid, bool)
+          : grid{grid}, intp_method{InterpolationMethod()},
+            grid_rational{GridRational()}, search_method{SearchMethod()} {}
+
+      /**
+       * The general procedure of the interpolatorar is to initialize the
+       * ressources for the interpolation method on a grid and test if the error
+       * is within a bound. If this is not the case the interpolator will refine
+       * the grid and repeat its procedure.
+       */
+      void initialize_iteratively() {
+        this->compute_grid_error();
+        while (this->grid_error > this->error_bound &&
+               this->grid.size() < this->max_grid_points) {
+          this->grid_fineness++;
+          this->compute_grid_error();
+        }
+      }
+
+      virtual void compute_grid_error() = 0;
+
+      // The boundary points of the range of interpolation
+      const double x1{0.};
+      const double x2{0.};
+      // grid in the range [x1,x2]
+      Vector_t grid{};
+      const double error_bound{1e-5};
+      double grid_error{0.};
+
+      const int max_grid_points{10000000};  // 1e7
+      int grid_fineness{5};
```

Comment:
Those mean qualitatively different things. Precision not only strongly implies precision in the _computed_ value (rather than simply the fineness of the grid spacings), but also seems to require a numerical value (this seems more like a "level of fineness", order-of-magnitude sort of parameter)

Reply:
fineness sounds weird though, do you have a better name?

Reply:
no, and I'm pretty sure it's quite standard, with approximately the same technical meaning across a variety of disciplines: https://www.google.com/search?client=firefox-b-d&q=grid+fineness . Probably it only sounds weird if you understand French well enough that it looks like a misspelling of "finesse"... but both are valid words in English, and with about the same level of usage: https://www.oed.com/view/Entry/70381?redirectedFrom=fineness#eid (esp. sense 3)

Reply:
Ok 👍